pkg/state/transaction_checker.go

```diff
@@ -1588,23 +1761,47 @@ func (tc *transactionChecker) checkCommitToGenerationWithProofs(
 		return txCheckerData{}, fmt.Errorf("invalid NextGenerationPeriodStart: expected %d, got %d",
 			nextPeriodStart, tx.GenerationPeriodStart)
 	}
+	// Calculate sender's address.
+	addr, adErr := proto.NewAddressFromPublicKey(tc.settings.AddressSchemeCharacter, tx.SenderPK)
+	if adErr != nil {
+		return txCheckerData{}, fmt.Errorf("failed to get address from public key: %w", adErr)
+	}
 	// Check that the sender has no other CommitToGeneration transaction with the same nextGenerationPeriodStart.
 	exist, err := tc.stor.commitments.newestExists(tx.GenerationPeriodStart, tx.SenderPK, tx.EndorserPublicKey)
 	if err != nil {
 		return txCheckerData{}, errors.Wrap(err, "failed to check existing commitment for the sender")
 	}
 	if exist {
-		addr, adErr := proto.NewAddressFromPublicKey(tc.settings.AddressSchemeCharacter, tx.SenderPK)
-		if adErr != nil {
-			return txCheckerData{}, stderrors.Join(
-				fmt.Errorf("generator %q has already committed to the next period %d",
-					tx.SenderPK.String(), nextPeriodStart),
-				errors.Wrap(adErr, "failed to get address from public key"),
-			)
-		}
 		return txCheckerData{}, fmt.Errorf("generator %q has already committed to the next period %d",
 			addr, nextPeriodStart)
 	}
+	// Check that the sender has enough generating balance.
+	gb, err := tc.stor.balances.newestGeneratingBalance(addr.ID(), blockHeight)
+	if err != nil {
+		return txCheckerData{}, errors.Wrap(err, "failed to get sender's generating balance")
+	}
+	mgb := tc.stor.features.minimalGeneratingBalanceAtHeight(blockHeight, info.currentTimestamp)
+	fee, err := safecast.Convert[int64](tx.Fee)
+	if err != nil {
+		return txCheckerData{}, fmt.Errorf("failed to convert fee to int64: %w", err)
+	}
+	futureSpends, err := common.AddInt[int64](Deposit, fee)
+	if err != nil {
+		return txCheckerData{}, fmt.Errorf("failed to calculate future spends: %w", err)
+	}
+	futureSpendsUint64, err := safecast.Convert[uint64](futureSpends)
+	if err != nil {
+		return txCheckerData{}, fmt.Errorf("failed to convert future spends to uint64: %w", err)
+	}
+	balanceAfter, err := common.SubInt[uint64](gb, futureSpendsUint64)
+	if err != nil {
+		return txCheckerData{}, fmt.Errorf("failed to calculate balance after transaction application: %w", err)
+	}
```

Comment:
What if user has commitment for current period? How it affects generation balance?

Reply:
Generation balance calculated as minimal Effective balance for the last 1000 blocks. Effective balance in turn calculated like this:
```golang
...
// Effective balance is calculated as: Balance + LeaseIn - LeaseOut - Deposit.
...
```
https://github.com/wavesplatform/gowaves/blob/d5ffa83ceccff8f0cd919e362c0822a6a682d310/pkg/state/balances.go#L49
So, the deposit for the current generation period is subtracted from the generation balance.